### DISCUSSION TITLE
Symfony 5.3 - Can't display login error messages

### DISCUSSION THREAD

**Discussion by ploegsteert** [symfony/symfony#42386] [Q&A] [answered]
Hello,

I used to authenticate users with the guard method, was working perfectly.

Now I try to edit my code to use the [new authenticator-based system](https://symfony.com/doc/current/security/authenticator_manager.html) as the old one is deprecated, everything is working fine except for the error messages.

If wrong credentials, an exception is thrown, I can dump it in the onAuthenticationFailure() method but can't retrieve it from controller with the $authenticationUtils->getLastAuthenticationError() method.

Could anyone help me with this please ? The online documentation for this new auth system is quite short for the moment.

**/src/Security/LoginFormAuthenticator.php**

```php
    <?php
    
    namespace App\Security;
    
    use Symfony\Component\Security\Http\Authenticator\AbstractAuthenticator;
    
    use Symfony\Component\HttpFoundation\Request;
    use Symfony\Component\HttpFoundation\Response;
    use Symfony\Component\HttpFoundation\RedirectResponse;
    
    use Symfony\Component\Security\Http\Authenticator\Passport\Passport;
    use Symfony\Component\Security\Http\Authenticator\Passport\Credentials\PasswordCredentials;
    use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
    use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
    use Symfony\Component\Security\Http\Authenticator\Passport\Badge\CsrfTokenBadge;
    use Symfony\Component\Security\Http\Authenticator\Passport\Badge\RememberMeBadge;
    
    use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
    use Symfony\Component\Security\Core\Exception\AuthenticationException;
    use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
    
    use Symfony\Component\Security\Http\Util\TargetPathTrait;
    use Doctrine\ORM\EntityManagerInterface;
    use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
    use Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface;
    
    use App\Entity\Utilisateur;
    use Symfony\Component\Security\Core\Security; 
    
    //todelete
    use Symfony\Component\Security\Http\Authentication\AuthenticationUtils;
    
    
    
    class LoginFormAuthenticator extends AbstractAuthenticator
    {
    	
    	use TargetPathTrait;
    	
    	public const LOGIN_ROUTE = 'app_login';
    	
        private $urlGenerator;
    
    	public function __construct(UrlGeneratorInterface $urlGenerator)
    	{
            $this->urlGenerator = $urlGenerator;
            
    	}
    	
    	public function supports(Request $request): ?bool
    	{
    		return self::LOGIN_ROUTE === $request->attributes->get('_route')
                && $request->isMethod('POST');
    	}
    	
    	public function authenticate(Request $request): PassportInterface
    	{
    		$credentials = [
                'email' => $request->request->get('email'),
                'password' => $request->request->get('password'),
                'csrf_token' => $request->request->get('_csrf_token'),
            ];
            
            $request->getSession()->set(
                Security::LAST_USERNAME,
                $credentials['email']
            );
    		
    		return new Passport(
    				new UserBadge($credentials['email']),
    				new PasswordCredentials($credentials['password']),
    				[
    					new CsrfTokenBadge('login', $credentials['csrf_token']),
    					new RememberMeBadge()
    				]
    			);
    		
    	}
        
        public function onAuthenticationSuccess(Request $request, TokenInterface $token, string $firewallName): ?Response
        {
        	if ($targetPath = $this->getTargetPath($request->getSession(), $firewallName)) {
                return new RedirectResponse($targetPath);
            }
            return new RedirectResponse($this->urlGenerator->generate('index'));
        }
        
        public function onAuthenticationFailure(Request $request, AuthenticationException $exception): ?Response
        {
        	return new RedirectResponse($this->urlGenerator->generate(self::LOGIN_ROUTE));
        }
    }
```

**/src/Controller/SecurityController.php**

```php
    <?php
    
    namespace App\Controller;
    
    use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
    use Symfony\Component\HttpFoundation\Response;
    use Symfony\Component\Routing\Annotation\Route;
    use Symfony\Component\Security\Http\Authentication\AuthenticationUtils;
    
    class SecurityController extends AbstractController
    {
        /**
         * @Route("/login", name="app_login")
         */
        public function login(AuthenticationUtils $authenticationUtils): Response
        {
            // get the login error if there is one
            $error = $authenticationUtils->getLastAuthenticationError();
            // last username entered by the user
            $lastUsername = $authenticationUtils->getLastUsername();
    
            return $this->render('login.html.twig', ['last_username' => $lastUsername, 'error' => $error]);
        }
    
        /**
         * @Route("/logout", name="app_logout")
         */
        public function logout(): void
        {
            throw new \LogicException('This method can be blank - it will be intercepted by the logout key on your firewall.');
        }
    }
```

**/config/packages/security.yaml**

```yml
    security:
        enable_authenticator_manager: true
        password_hashers:
            App\Entity\Utilisateur:
                algorithm: auto
    
        # https://symfony.com/doc/current/security.html#where-do-users-come-from-user-providers
        providers:
            # used to reload user from session & other features (e.g. switch_user)
            app_user_provider:
                entity:
                    class: App\Entity\Utilisateur
                    property: email
        firewalls:
            dev:
                pattern: ^/(_(profiler|wdt)|css|images|js)/
                security: false
            main:
                #anonymous: true
                lazy: true
                provider: app_user_provider
                
                form_login: ~
                #entry_point: form_login
                custom_authenticators:
                  - App\Security\LoginFormAuthenticator
                #guard:
                  #authenticators:
                    #- App\Security\LoginFormAuthenticator
                logout:
                  path: app_logout
                remember_me:
                  secret: '%kernel.secret%'
                  lifetime: 31536000
                  path: /
    
                # activate different ways to authenticate
                # https://symfony.com/doc/current/security.html#firewalls-authentication
    
                # https://symfony.com/doc/current/security/impersonating_user.html
                # switch_user: true
    
        # Easy way to control access for large sections of your site
        # Note: Only the *first* access control that matches will be used
        access_control:
             - { path: ^/login, roles: PUBLIC_ACCESS }
             - { path: ^/, roles: ROLE_USER }
            # - { path: ^/profile, roles: ROLE_USER }

**Comment by lyrixx** [comment, marked as answer]
Hello, 

I recommend you to use the native form login authenticator (btw you are using it **and** the authenticator, that's a bit suspicious)

If you don't want, you can generate one with the maker bundle to see how it works.

Anyway the first glance, I don't find the typo. Maybe because of both authenticator

**Reply by ploegsteert** [reply]
Thanks for your answer.

I'm not sure to understand everything.

According to the [doc](https://symfony.com/doc/current/security.html) :

"But, in most cases - including a login form - we recommend building a Guard Authenticator" but just below : "Deprecated since version 5.3:Guard authenticators are deprecated since Symfony 5.3 in favor of the new authenticator-based system."

I used to follow this [guide](https://symfony.com/doc/current/security/form_login_setup.html), but had deprecations when updated from 5.2 to 5.3.

**Reply by lyrixx** [reply]
The documentation is a bit outdated indeed.

To summarize: 

* if a builtin authenticator exist => it's better to use it
* if not, use a custom authenticator
* on previous version of symfony, one can use a guard instead of an authenticator. But that's deprecated.

---

But if you want to stick with a custom authenticator, you must add the following line 
```
$request->getSession()->set(Security::AUTHENTICATION_ERROR, $exception);
```
in the `onAuthenticationFailure` method.

You can also inject the `Symfony\Component\Security\Http\Authentication\DefaultAuthenticationFailureHandler` and use it

**Reply by ploegsteert** [reply]
Ok thanks to your answers and some on stackoverflow, I managed to get the login and error messages working by extending AbstractLoginFormAuthenticator which has a onAuthFailure method which write to the session.

Just one more thing : "if a builtin authenticator exist => it's better to use it" is there a doc page where we can find natives authenticators or implement it ?

Thank you so much for your help

**Reply by lyrixx** [reply]
I guess this is the best page https://symfony.com/doc/current/security/auth_providers.html
(I don't usually browse the doc)

You can also run this commande:
```
bin/console config:dump security
```

**Reply by ploegsteert** [reply]
Thank you for everything

**Reply by ahundiak** [reply]
I was also confused about the builtin authenticator advice.  form_login is also deprecated and is going away.  It's a good thing that you completely removed it from security.yaml.

**Reply by lyrixx** [reply]
No it's not. 
Everything will be clarified soon, once this pr get merged 
https://github.com/symfony/symfony-docs/pull/15503

**Reply by ahundiak** [reply]
Hmmm.  I think I will adopt a wait and see approach.  The [blog article](https://symfony.com/blog/new-in-symfony-5-3-guard-component-deprecation) make it clear that form_login is going away and I can't see any reasonable reason to support two authentication processes going forward.  But time will tell I suppose.

**Reply by wouterj** [reply]
All built-in authentication providers have been rewritten to the new authenticator implementations. So form_login, login_link, json_login, http_basic, remember_me, x509 and remote_user will not be removed.

**Reply by ahundiak** [reply]
Fair enough.  I stand corrected.  Thanks.

Update: Thought about this some more and it seems a bit sad that they will not be removed.  I was hoping that the new authentication scheme might simplify might configuration.  If you want authentication then you make an authentication class and add it to your firewall.  Pretty straight forward.  Maybe update the make:auth command and add json/ldap etc options.  All of that extra firewall config functionality could simply go away.

The way things stand now a developer can go the the security docs and see that they should use make:user to create a user and configure a provider and a password hasher if applicable.  Easy enough.

Then they get to the authentication docs and off they go into la la land.  I have been using Symfony 2.0 and even wrote some of my pre-guard authenticators but I can't figure out what the docs are trying to say.  And I see countless questions on stackoverflow.

If the only choice was to use make:auth to create an authenticator class then hey, huge sections of config could just disappear for 6.0.  Just a thought.

**Comment by ploegsteert** [comment]
Ok so I updated my security.yaml according to your answer :

```yml
security:
    enable_authenticator_manager: true
    password_hashers:
        App\Entity\Utilisateur:
            algorithm: auto

    providers:
        app_user_provider:
            entity:
                class: App\Entity\Utilisateur
                property: email
    firewalls:
        dev:
            pattern: ^/(_(profiler|wdt)|css|images|js)/
            security: false
        main:
            lazy: true
            provider: app_user_provider
            
            entry_point: App\Security\LoginFormAuthenticator
            custom_authenticators:
              - App\Security\LoginFormAuthenticator
            logout:
              path: app_logout
            remember_me:
              secret: '%kernel.secret%'
              lifetime: 31536000
              path: /
```
So the login still working as excepted, except for errors wich are not retrieved with the $authenticationUtils->getLastAuthenticationError() method... Is there another method to get Auth Exceptions ?

**Reply by lyrixx** [reply]
See https://github.com/symfony/symfony/discussions/42386#discussioncomment-1133697 :)
